src/interpreter/explicitControlEvaluator.ts

```diff
@@ -1,7 +1,286 @@
+      state.agenda.push(command.arguments[index]);
+    }
+    state.agenda.push(command.id);
+  },
+  Constant: (command: Constant, state: ExplicitControlEvaluatorState) => {
+    state.stash.push(command);
+  },
+  ContinueStatement: (
+    command: ContinueStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  DefaultStatement: (
+    command: DefaultStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  DoWhileStatement: (
+    command: DoWhileStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  EmptyStatement: (
+    command: EmptyStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  Environment: (
+    command: EnvironmentInstr,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  ExpressionStatement: (
+    command: ExpressionStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  ForStatement: (
+    command: ForStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  FunctionApplication: (
+    command: FunctionApplicationInstr,
+    state: ExplicitControlEvaluatorState
+  ) => {
+    // Args are in order of left to right
+    const args: Value[] = [];
+    for (let index = 0; index < command.numOfArgs; index++) {
+      args.push(state.stash.pop());
+    }
+
+    const functionAddress = getEnvironmentValue(
+      command.functionId.name,
+      state.environment
+    );
+    const closureIdx = state.memory.get(functionAddress);
+    const closure = state.textMemory.get(closureIdx);
+
+    if (closure.params.length !== args.length) {
+      throw new InvalidFunctionApplication(
+        `Function takes in ${closure.params.length} arguments but ${args.length} arguments were passed in`
+      );
+    }
+
+    if (isEmptyStatement(closure.body)) {
+      return;
+    }
+
+    // TODO: Handle Tail Call in future.
+    if (
+      state.agenda.size() === 0 ||
+      state.agenda.peek().type === 'Environment'
+    ) {
+      // Don't need current environment, push FunctionMarkInstr only and not EnvironmentInstr
+      state.agenda.push(constructFunctionMarkInstr());
+    } else {
+      state.agenda.push(constructEnvironmentInstr(state.environment));
+      state.agenda.push(constructFunctionMarkInstr());
+    }
+
+    // Params and block variables should be in the same scope
+    const paramNames = closure.params.map((param) => param.name);
+    const paramAddresses = state.memory.stackFunctionCallAllocate(
+      getArgNumbers(args)
+    );
+    const blockVariableDeclarations = getBlockVariableDeclarationNames(
+      closure.body
+    );
+    const blockVariableAddresses = allocateStackAddresses(
+      blockVariableDeclarations.length,
+      state.memory
+    );
+
+    state.environment = extendEnvironment(
+      [...paramNames, ...blockVariableDeclarations],
+      [...paramAddresses, ...blockVariableAddresses],
+      closure.environment
+    );
+
+    for (let i = closure.body.body.length - 1; i >= 0; i--) {
```

Comment:
`.body.body` is a bit of a code smell :/

Can `body` in `BlockStatement` be `items` instead since it is an array of `BlockItem`s?